[FIXED JENKINS-28840] Deadlock between Queue.maintain and Executor.interrupt

stephenc opened the pull request:
See [JENKINS-28840](https://issues.jenkins-ci.org/browse/JENKINS-28840)

More fun here:
- All this originates from `Executor` extending `Thread`.
- There is funky logic in the lock handling code of the JVM that makes assumptions
  about how it might proceed with the lock when the thread holding the lock has its
  interrupt flag set.
- Really it would be better if `Executor` did not extend `Thread` as that way we wouldn't
  have to deal with some of that complexity. But OTOH we are where we are and backwards
  compatibility may make such a change not possible without a lot of breakage.
- Fixing the issue at hand, firstly requires that interrupting a `Computer` happens with the
  `Queue` lock held (to speed up tests we have `Jenkins.cleanUp` get the lock for all `Computer`s)
  That prevents the `Queue` maintain thread from getting caught
- Secondly, when removing an executor from a computer we process the removal while
  holding the `Queue` lock, but we move the removal itself to a separate thread if we cannot
  get the `Queue` lock in order to avoid deadlock.
- Also add helper methods to wrap tasks to be performed while holding the lock
  and a helper method for `Runnable`s that exposes the `tryLock` functionality

@olivergondza hopefully this fixes the issue and it's not that I didn't run the flaky test case enough times on my machine. For back-porting to 1.609.2 we should obviously mark the new API methods in `Queue` as `@Restricted`

@reviewbybees

stephenc committed:
[FIXED JENKINS-28840] Deadlock between Queue.maintain and Executor.interrupt

More fun here:

- All this originates from Executor extending Thread.
- There is funky logic in the lock handling code of the JVM that makes assumptions
  about how it might proceed with the lock when the thread holding the lock has its
  interrupt flag set.
- Really it would be better if Executor did not extend Thread as that way we wouldn't
  have to deal with some of that complexity. But OTOH we are where we are and backwards
  compatibility may make such a change not possible without a lot of breakage.
- Fixing the issue at hand, firstly requires that interrupting a Computer happens with the
  Queue lock held (to speed up tests we have Jenkins.cleanup get the lock for all Computers)
  That prevents the Queue maintain thread from getting caught
- Secondly, when removing an executor from a computer we process the removal while
  holding the Queue lock, but we move the removal itself to a separate thread if we cannot
  get the Queue lock in order to avoid deadlock.
- Also add helper methods to wrap tasks to be performed while holding the lock
  and a helper method for Runnables that exposes the tryLock functionality

rsandell commented:
:+1:

olivergondza commented:
@stephenc, did you manage to reproduce the deadlock without the fix? As I reported it did not happen every time.

oleg-nenashev commented on core/src/main/java/hudson/model/Queue.java:
```diff
@@ -1219,6 +1219,59 @@ public static void withLock(Runnable runnable) {
+    public static boolean tryWithLock(Runnable runnable) {
+        final Jenkins jenkins = Jenkins.getInstance();
+        final Queue queue = jenkins == null ? null : jenkins.getQueue();
+        if (queue == null) {
```
NIT: it should not really happen (?). Otherwise, we still have a risk of surprising behaviors on Jenkins startup

oleg-nenashev replied:
BTW, I would always use an external lock to keep the consistency

oleg-nenashev commented:
There's a risk of new regressions and deadlocks, but :+1: for merge in order to test it as much as possible before LTS backporting timeslot ends.

jglick commented:
`CommandLauncherTest.commandFails` is probably unrelated but should be evaluated.

jglick commented on core/src/main/java/hudson/model/Computer.java:
```diff
@@ -990,14 +990,13 @@ public final long getDemandStartMilliseconds() {
                 synchronized (Computer.this) {
                     executors.remove(e);
                     addNewExecutorIfNecessary();
-                    if (!isAlive()) // TODO except from interrupt/doYank this is called while the executor still isActive(), so how could !this.isAlive()?
```
Is there a reason you deleted my comment?

stephenc replied:
Yes as this call can now happen on the background thread so it may now be not alive when the thread scheduler gets around to scheduling execution

jglick commented on core/src/main/java/hudson/model/Queue.java:
```diff
@@ -1219,6 +1219,59 @@ public static void withLock(Runnable runnable) {
     }
 
+    /**
+     * Some operations require to be performed with the {@link Queue} lock held. Use one of these methods rather
```
grammar

jglick commented on core/src/main/java/hudson/model/Queue.java:
```diff
@@ -1219,6 +1219,59 @@ public static void withLock(Runnable runnable) {
+     * Some operations require to be performed with the {@link Queue} lock held. Use one of these methods rather
+     * than locking directly on Queue in order to allow for future refactoring.
+     * @param runnable the operation to perform.
+     * @return {@code true} if the lock available and the operation performed. 
```
grammar

jglick commented on core/src/main/java/hudson/model/Queue.java:
```diff
@@ -1247,6 +1300,26 @@ protected void _withLock(Runnable runnable) {
+     * @return {@code true} if the lock available and the operation performed. 
+     * @since 1.FIXME
+     */
+    protected boolean _tryWithLock(Runnable runnable) {
```
Why is this `protected`? `Queue` is not intended to be subclassed.

jglick commented:
:+1: without fully understanding the fix, but sounds reasonable.

stephenc commented:
@olivergondza over 2 hours of running

while true; do mvn surefire:test ; done

on the unpatched code and I have not got a deadlock, so no. I have not been able to reproduce your failure, but this should fix it.

stephenc committed:
Test case was subject to timing bugs

- I suspect the synchronization on Jenkins is a bug also... but not causing the test failure, so will ignore for now

stephenc committed:
prepare for merge

stephenc committed:
Merge remote-tracking branch 'jenkins/master' into jenkins-28840
